cim-server/src/main/java/com/crossoverjie/cim/server/server/CIMServer.java

```diff
@@ -97,4 +116,55 @@ public void sendMsg(SendMsgReqVO sendMsgReqVO){
         future.addListener((ChannelFutureListener) channelFuture ->
                 log.info("server push msg:[{}]", sendMsgReqVO.toString()));
     }
+
+    @RedisLock(key = "T(java.lang.String).format('lock:offlineMsg:%s', #userId)",
+            waitTime = 5, leaseTime = 30)
+    public void sendOfflineMsgs(Long userId) {
+        NioSocketChannel channel = SessionSocketHolder.get(userId);
+        if (channel == null) {
+            log.error("client {} offline!", userId);
+            return;
+        }
+
+        List<OfflineMsg> fetchMsgs = offlineMsgStore.fetch(userId);
+        if (fetchMsgs.isEmpty()) {
+            return;
+        }
+        fetchMsgs.sort(Comparator.comparing(OfflineMsg::getCreatedAt));
+
+        int batchSize = fetchMsgs.size();
+        ChannelFuture lastWriteFuture = null;
+        for (OfflineMsg offlineMsg : fetchMsgs) {
+            log.info("offlineMsg:{}", offlineMsg);
+            Request protocol = Request.newBuilder()
+                    .setRequestId(offlineMsg.getMessageId())
+                    .setReqMsg(offlineMsg.getContent())
+                    .putAllProperties(offlineMsg.getProperties())
+                    .setCmd(BaseCommand.MESSAGE)
+                    .build();
+
+            lastWriteFuture = channel.write(protocol);
+        }
+
+        channel.flush();
+
+        if (lastWriteFuture != null) {
+            lastWriteFuture.addListener((ChannelFutureListener) future -> {
```

Comment:
It is recommended to reuse `com.crossoverjie.cim.server.server.CIMServer#sendMsg` function (just add a new parameter `cmd`).